jicoco/src/main/java/org/jitsi/xmpp/mucclient/MucClient.java

```diff
@@ -857,4 +860,45 @@ private void resetLastPresenceSent()
             lastPresenceSent = null;
         }
     }
+
+    /**
+     * Detects ping failures and in case of two consecutive failures where the connection is still
+     * connected and authenticated we will close it and let the connectRetry task handle that.
+     */
+    private class PingFailedListenerImpl
+        implements PingFailedListener
+    {
+        @Override
+        public void pingFailed()
+        {
+            logger.warn("XMPP Ping failed");
+
+            boolean res = false;
+            try
+            {
+
+                PingManager pingManager = PingManager.getInstanceFor(xmppConnection);
+                if (pingManager != null)
+                {
+                    res = pingManager.pingMyServer(false);
```

Comment:
So this isn't waiting for 2 normal ping failures, this is waiting for one ping failure and then immediately sending another ping to see if that succeeds, is that right?

Reply:
Yep.

Reply:
Should we instead wait for 2 failures at the regular ping interval?

Reply:
So we cannot count on the ping manager, as at this point it had stopped. But we can schedule a task that will do the second ping after the configured timeout. That will give a little more time, another 30 seconds to the link ... 
Do you want me to add that? I have no opinion which one is better ...

Reply:
Ah ha!  Ok, I didn't know that.  So the PingManager stops after a single failure?  I think it would be good to add a comment to that effect (that we must do our own ping here since PingManager has stopped).   But besides that, yeah I think whatever makes more sense--2 immediate ping failures or if we want to give some buffer if there was a hiccup.

Reply:
Ok, I've added a comment and will wait for @bgrozev to take a look and give his opinion.